File: ruby/ch1/ex123.rb
module Ex121

def divides(a, b)
  0 == b % a
end

def find_divisor(n, test)
  if test * test > n
    n
  elsif divides(test, n)
    test
  else
    find_divisor(n, next_num(test))
  end
end 

def next_num(n)
  n == 2 ? 3 : n + 2
end

def smallest_divisor(n)
  find_divisor(n, 2)
end

end


include Ex121
print "Testing\n"
print smallest_divisor(199) == 199 , "\n"
print smallest_divisor(1999) == 1999 , "\n"
print smallest_divisor(19999) == 7 , "\n"

print "done", "\n"

File: scala/src/ch1/ex123.scala
// run with> scala exXXX.scala

def divides(a: Int, b: Int): Boolean = b % a == 0

def smallest_divisor(n: Int): Int = find_divisor(n, 2)

def find_divisor(n: Int, test: Int): Int = test match {
  case x if x * x > n => n
  case x if divides(test, n) => test
  case _ => find_divisor(n, next(test))
}

def next(n: Int): Int = n match {
  case 2 => 3
  case _ => n + 2
}

assert(smallest_divisor(199) == 199)
assert(smallest_divisor(1999) == 1999)
assert(smallest_divisor(19999) == 7)

println("done")
